libraries/Microsoft.Bot.Builder.Dialogs/Memory/Scopes/BotStateMemoryScope.cs

```diff
@@ -40,7 +37,9 @@ public override object GetMemory(DialogContext dialogContext)
                 throw new ArgumentNullException($"{nameof(dialogContext)} is null");
             }
 
-            var cachedState = dialogContext.Context.TurnState.Get<object>(typeof(T).Name);
+            var botState = GetBotState(dialogContext);
+            var contextServiceKey = typeof(BotState).GetField("_contextServiceKey", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(botState);
+            var cachedState = dialogContext.Context.TurnState.Get<object>((string)contextServiceKey);
             return cachedState.GetType().GetProperty("State").GetValue(cachedState);
```

Comment:
What? This seems really goofy, why isn't this just:
            var botState = dialogContext.Context.TurnState.Get<T>();
            if (botState != null)
            {
                return botState.Get(dialogContext.Context);
            }

Reply:
Great question, @tomlm, and thank you kindly for the review. Yes it does seem goofy, and that's why my first "best" solution was to get rid of the reflection and just call `BotState.Get` like you've said. You can still see that solution in some of the commits in this PR, before I reverted back to using reflection.

I discovered that the reason we can't use `BotState.Get` is because it unfortunately returns a copy of the cached turn state instead of the actual turn state, as explained in the method's documentation. When I tried using `BotState.Get` I found that many tests were failing, and this was because the adaptive dialogs were trying to modify the state returned by `BotStateMemoryScope<T>.GetMemory`. If we return a copy of the actual turn state then we'll return a different object every time and any changes we make to it will not be reflected in the objects returned by other calls. So for example when `DialogStateManager` originally gets its user state object and stores it in its `Values` property and then tries to set `user.counter` it will get a new user state object and set `user.counter` in that and the user state in its `Values` property will be unaffected and the change will be lost.

I hope the explanation I gave in the original comment of this PR makes sense too. I described four potential solutions to this problem and explained why I went with the solution I did. If you would like to go with a different solution or if you can think of a fifth solution, please let me know.

Reply:
@tomlm, have you had a chance to look at this with Kyle?